### ISSUE TITLE
Clarification about SOMD parameters

### ISSUE THREAD

**Issue by msuruzhon** [michellab/BioSimSpace#258]
Hi,

I have been trying to run some SOMD simulations and I realised I don't understand what some of the parameters mean, even after reading the part of the code that's responsible for setting them. In particular, I don't understand what `ncycles`, `nmoves`, `ncycles_per_snap` and `buffered coordinates frequency` are supposed to do. Also, how do these parameters relate to the `report_interval` and `restart_interval`? Judging by the AMBER code, the `restart_interval` should set the number of steps that the engine has to wait before it outputs restart files or coordinates, while `report_interval` is the number of steps between energy evaluations. How does that translate in SOMD?

Many thanks.

**Comment by lohedges** [comment]
Hi @msuruzhon. SOMD is different to the other engines that we support in that it was originally designed to allow the execution of various types of _move_, each with a different weight. A move could be something like a regular MD step, a box transformation, etc. This would allow the user to mix different types of move to aid equilibration, or to do some different type of sampling according to some MCMC scheme. In terms of the parameters in the config file, these mean:

* `nycles`: The total number of "sampling" cycles.
* `nmoves`: The number of moves that are performed be cycle. (Could be different moves, but here we only use repeated MD steps.)
* `ncycles_per_snap`: The number of cycles that are performed between saving snapshots to the trajectory file.
* `buffered_coordinates_frequency`: The number of time steps between saving of coordinates during a cycle of MD. 0 disables buffering.

To see more info about all of the supported options you can run:

```
somd -H
```

or

```
somd-freenrg -H
```

(Just to note that some of the of the defaults that are reported are actually incorrect, e.g. the units can be wrong.)

With the above move and cycle scheme I had to do some some work in order to make the reporting consistent with how it is implemented for the other engines, e.g. by using the `restart_interval` and `report_interval` options. Your interpretation of these is absolutely correct, i.e. time between saving a trajectory snapshot and restart file and time between energy evaluations and logging. The issue with applying these to SOMD is that you can only get certain information at the cycle frequency, and it is most efficient to have few long cycles since it to ensure that you are running on the GPU for longer. (At the end of a cycle information is copied to the host, converted to a Sire object, then back again. Doing this frequently is slow.) Another wrinkle with SOMD is that some of the parameters in the config file have (undocumented) constraints on their values, and in some cases they are actually coupled, e.g. `buffered_coordinates_frequency`, which must be a multiple of the `energy_frequency`. Making things self consistent means that the `report_interval` and `restart_interval` are interpreted _slightly_ differently with SOMD, i.e. I keep things as close as I can within the constraints, but adjust the values as necessary. (Things become more accurate for longer simulations.)

For AMBER you should just be able to use the two options in the normal way, i.e. as I have done for the other protocols that are supported with AMBER. For the free energy protocol (your `FreeEnergyProduction`) we have recently adjusted the defaults for these parameters since such simulations will typically be long GPU runs, i.e. it makes sense to make the intervals much larger.

I hope this helps. In summary: SOMD is a bit of an outlier in respect to the other MD engines, since it was originally designed to work in a quite different way. Just interpret the `restart_interval` and `report_interval` as you have done, making the intervals consistent with how they are implemented for, e.g. GROMACS.

**Comment by msuruzhon** [comment]
Many thanks for the detailed anwer @lohedges. I am still a bit lost on why we don't set the default values for SOMD as follows:

```python
ncycles = nsteps // nmoves                                     # For a total of ncycles * nmoves timesteps
nmoves = min(restart_interval, report_interval)             
ncycles_per_snap = max(1, restart_interval // nmoves)
buffered_coordinates_frequency = ((nmoves * ncycles_per_snap) // energy_frequency) * energy_frequency
energy_frequency = report_interval                             # Same behaviour as AMBER
```
The main differences from the current implementation are setting the `energy_frequency` to be equal to the `report_interval`, rather than `250` (is this 250 timesteps, by the way?, i.e. 0.5 ps?), as well as setting the `buffered_coordinates_frequency` to be equal to `nmoves * ncycles_per_snap` and then making sure this is divisible by the `energy_frequency=report_interval`. Would that not work? From what I can see this allows us to set the `energy_frequency` in a way that's similar to the behaviour in the other MD engines.

**Comment by lohedges** [comment]
Yes, the `energy_frequency` is 250 timesteps (moves). As far as I understand, this parameter sets the frequency at which the free energy gradients are computed, so directly affects the sampling, i.e. it doesn't determine the frequency at which results are reported. If we can get away with increasing it to match the report interval as suggested, then I'd be happy with standardising things. I think I did write things as you suggest when I first wrote the SOMD config writer, but then adjusted it to the current form for consistency with how things were previously done.

Perhaps @jmichel80 has more input on the exact requirements for the `energy_frequency` option. (For reference, the default is 1, and we now agree with the value that Cresset use. Previously we used 100.)

**Comment by jmichel80** [comment]
energy_frequency should be at least 100, otherwise you will have high overheads as the simulation pauses MD integration frequently to swap Hamiltonians. You will also have smaller simfile.dat output and faster MBAR estimation.

**Comment by msuruzhon** [comment]
Thanks @lohedges and @jmichel80. Does that mean that we can simply set `energy_frequency = report_interval` then? By the way, I just saw that the current defaults for the `report_interval` are 200000 steps. Isn't this value too high? This will result in only one MBAR sample per 0.4 ns if you use a 2 fs timestep. Maybe we can set this default to some value between 500 and 5000 (let's say 500) and then this value will also make sense for `energy_frequency`?

**Comment by lohedges** [comment]
Thanks, @jmichel80, agreed. How large would we be able to push things? The current defaults are tuned to give few long cycles so that we run on the GPU for as long as possible. We have:

* `runtime = 4ns`
* `timestep = 2fs`
* `report_interval = 20000`

This would give a simulation of 10 cycles, so I assume that we would only get 10 energy samples, which would be hugely insufficient from a sampling perspective.

**Comment by lohedges** [comment]
> I just saw that the current defaults for the report_interval are 200000 steps. Isn't this value too high? This will result in only one MBAR sample per 0.4 ns if you use a 2 fs timestep

No, we are sampling at the `energy_frequency`, i.e. one sample per 250 steps. We are only _writing_ at the `report_interval`.

I think there is just some confusion as to how SOMD works and what the parameters mean. SOMD only writes anything on a per cycle basis, so the `report_interval` takes this meaning. (Even though SOMD doesn't really _write_ anything to log files like the other engines do.

As you say, setting the other FEP engine settings on the basis of SOMDs will result in undersampling if they can only sample and report at the same time, i.e. not buffer the data. Given that SOMD doesn't actually do any different type of move it would be nice to just re-write it to do one long cycle with reporting performed in the same way as for the other engines.

**Comment by lohedges** [comment]
I think it would probably be easiest if we just accept that SOMD doesn't work like the other MD engines do, so use the `report_interval` and `restart_interval` as they are for the other engines, i.e. frequency of sampling energies and restarts, with the caveat that things won't be written to file as frequently as the user might expect, i.e. they are only being buffered at this frequency. I originally wrote the code to _write_ things at the expected frequency, but this results in lots of short MD cycles which gives very poor performance. I can always try to document the peculiarities of SOMD in the documentation.

We would also need to check (by running some test simulations) that we can actually adjust things as we desire. As mentioned, there are some constraints on the values that parameters can take so it's not always possible to adjust things as you expect. (In addition to the `energy_frequency` and `buffered_coordinates_frequency` coupling mentioned previously.)

**Comment by jmichel80** [comment]
For production settings I recommend 10 cycles for an overall 4ns sim. This will be hardly slower than  doing 1 cycle only. Benefits are that the simulation progress can be monitored, and if there is a crash it can be restarted from a restart file without having to start from scratch again. 
with ca. 10-20 lambda values in the schedule setting energy_frequency to 100 or 200 will not affect much performance and give you a few thousand data points for MBAR/TI analysis.

we will likely update default settings to run with a 4 fs timestep in early Q22

**Comment by lohedges** [comment]
That's exactly what we are currently doing :+1:

**Comment by msuruzhon** [comment]
@lohedges so 'reporting' in this setting means collecting all the generated energy values every 250 steps during the last 200000 steps and then dumping all of them at the end of these 200000 steps? Is this right?

If this is the case, then I am not sure how well these defaults translate into GROMACS, since GROMACS currently uses `nstenergy = report_interval`, which means (as far as I understand) that with the current setting only 10 samples will be written out in the `.xvg` file for BAR analysis with the default settings. Shouldn't we set these steps to 250 as well (meaning that `report_interval` is only used for log file output frequency)? If my understanding is correct, then that means that the `report_interval` parameter shouldn't be used in the AMBER free energy calculations at all?

Wouldn't it make more sense if the `report_interval` reflected the energy calculation frequency, instead of the write frequency (which we should be able to automatically set for SOMD behind the curtains in a more optimal way)? I would imagine that a user expects a new energy value every `report_interval` steps (I know I did).

Apologies for taking so long to wrap my head around this, for some reason I find it very confusing.

**Comment by lohedges** [comment]
> Apologies for taking so long to wrap my head around this, for some reason I find it very confusing.

No problem, it _is_ confusing! Definitely a case of trying to fit a square peg in a round hole. I wrote `Process.Somd` after some of the other engines and also got tripped up by SOMD's quirks.

> 'reporting' in this setting means collecting all the generated energy values every 250 steps during the last 200000 steps and then dumping all of them at the end of these 200000 steps? Is this right?

Yes, energies and trajectory frames are written after complete cycles, but are buffered in between according to the specified frequencies. For the current defaults this means that energies are written once per cycle, i.e. every `report_interval` integration steps, which is 20000, not 200000, i.e. there will be 800 energy samples per cycle, so 8000 in total.

> If this is the case, then I am not sure how well these defaults translate into GROMACS, since GROMACS currently uses nstenergy = report_interval, which means (as far as I understand) that with the current setting only 10 samples will be written out in the .xvg file for BAR analysis with the default settings.

In addition to the `nstenergy` option, GROMACS also has [nstcalenergy](https://manual.gromacs.org/documentation/2018/user-guide/mdp-options.html#mdp-nstcalcenergy), which specifies how frequently the energies are _calculated_. Although it will only write to file 10 times, the values will have been calculated and averaged over many more samples. (Unless I'm misunderstanding something.) In the GROMACS config I have set `nstcalenergy` to 250 steps to match the `energy_frequency` in SOMD.

AMBER uses `ntpr` to specify the frequency at which energies are written to file. There is also the `ntave` option that controls how frequently averages are written, but I've not used that. It also looks like there's an `ene_avg_sampling` option for PMEMD, but I'm not sure how this works in practice. In the manual I see the mention of a `bar_intervall` option to specify how frequently energies are sampled during FEP.

Perhaps we also need another option, such as `sample_frequency`, but I'm not sure this would be available in all engines. (Maybe it could just be used in the free energy protocol?) Reporting is kind of meaningless for SOMD, since, other than free-energy gradients for `somd-freenrg`, no other state information is reported.

Hope this makes sense. Possibly adding even more confusion ;-)
